src/lib.rs

```diff
@@ -117,17 +118,120 @@ pub fn select_coin_knapsack(
     inputs: &[OutputGroup],
     options: CoinSelectionOpt,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut adjusted_target = options.target_value
+        + options.min_drain_value
+        + calculate_fee(options.base_weight, options.target_feerate);
+    let mut smaller_coins = inputs
+        .iter()
+        .enumerate()
+        .filter(|&(index, output_group)| output_group.value < adjusted_target)
+        .map(|(index, output_group)| (index, *output_group))
+        .collect::<Vec<_>>();
+    // Sorting smaller_coins in descending order
+    smaller_coins.sort_by_key(|&(_, output_group)| Reverse(output_group.value));
+
+    knap_sack(adjusted_target, &smaller_coins, inputs, options)
 }
 
 /// adjusted_target should be target value plus estimated fee
 /// smaller_coins is a slice of pair where the usize refers to the index of the OutputGroup in the inputs given
 /// smaller_coins should be sorted in descending order based on the value of the OutputGroup, and every OutputGroup value should be less than adjusted_target
+fn knap_sack_calculate_accumulated_weight(
+    inputs: &[(usize, OutputGroup)],
+    selected_inputs: &HashSet<usize>,
+) -> u32 {
+    let mut accumulated_weight: u32 = 0;
+    for &(index, u) in inputs {
+        if selected_inputs.contains(&index) {
+            accumulated_weight += u.weight;
+        }
+    }
+    accumulated_weight
+}
 fn knap_sack(
     adjusted_target: u64,
     smaller_coins: &[(usize, OutputGroup)],
+    inputs: &[OutputGroup],
+    options: CoinSelectionOpt,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut selected_inputs: HashSet<(usize)> = HashSet::new();
+    let mut accumulated_value: u64 = 0;
+    let mut best_set: HashSet<(usize)> = HashSet::new();
+    // Assigning infinity to beginwith
+    let mut best_set_value: u64 = u64::MAX;
+    let mut rng = thread_rng();
+    for i in 1..=1000 {
+        for pass in 1..=2 {
+            for &(index, u) in smaller_coins {
+                //Simulate a coin toss
+                let toss_result: bool = rng.gen_bool(0.5);
+                let selected_element = (index, u);
```

Comment:
There is no need to club the index and OutputGroup together.
Use a better name like "coin" in place of "u"